rfcs/2023-radix-auth.md

```diff
@@ -0,0 +1,135 @@
+- Start Date: 2023-02-21
+- RFC PR: (leave this empty, to be filled in later)
+- Authors: Benoît Grélard
+
+# Radix Auth
+
+## Summary
+
+This RFC proposes adding a new project under the Radix UI umbrella (at the same level as "Radix Primitives", "Radix Colors", "Radix Icons"). This new product would be called "Radix Auth" and would provide a set of components specifically tailored to building great authentication experiences.
+
+It would build on top of the new `Form` primitive discussed in the [Radix Form primitive RFC](https://github.com/radix-ui/primitives/pull/1977) but provide higher level components geared towards specific scenarios such as sign-up, sign-in, password reset flows, etc.
+
+## Motivation
+
+At [WorkOS](https://workos.com), we build enterprise-ready features such as Single Sign-On, Directory-Sync and Audit Logs. A lot of these features revolve around great authentication experiences.
+
+We also build and maintain [Radix UI](https://radix-ui.com). We believe we could bring a similar level of quality to authentication experiences as people are used to with Radix primitives. Think everything you love about Radix: similar multi-parts API, great accessibility, great developer experience, etc.
+
+## Design
+
+At the most basic level, this would include new primitives such as `SignUp`, `SignIn` which would use the new [`Form` primitive](https://github.com/radix-ui/primitives/pull/1977) and provide an out of the box experience catered specifically for authentication flows. On top of that we would also provide other composable primitives such as `OTPInput`, `Password`, other primitives to visually represent password strength and more.
+
+### Examples
+
+#### A full sign-up UI composing a set of primitives
+
+https://user-images.githubusercontent.com/1539897/220597308-1fdc68a3-aaf5-46b2-a59d-4b5a1bc9ecaa.mp4
+
+---
+
+#### A password input with built-in show/hide functionality
+
+https://user-images.githubusercontent.com/1539897/220597112-cb3b74cc-e38b-4fdb-a0dc-83ca5de9c27a.mp4
+
+---
+
+#### An input for multi-factor authentication (one time password)
+
+https://user-images.githubusercontent.com/1539897/220597235-316c0251-0173-4c37-a522-f0b19cec447d.mp4
+
+---
+
+#### Other primitives to visually represent password strength
+
+https://user-images.githubusercontent.com/1539897/220597276-8896cf0f-7e8b-4d91-8c23-2a6a2ea866af.mp4
+
+Over time, we would keep providing value by incorporating learnings from our own products around single sign-on, multi-factor authentication, etc.
```

Comment:
What if we allow adding a custom function to calculate the strength of a password? Some products may force special characters; some may not. Others may require at least one capital letter, and some may not 🤔

We should have a standard strength calculation, but allowing overwriting of this calculation seems interesting to reach different requirements for each project.

Maybe something like:
## Donut
```jsx
<Donut.Root strengthEstimator={myCustomPasswordStrengthEstimator}>
   <Donut.Track />
   <Donut.Value />
</Donut.Root>
```

Here, my `myCustomPasswordStrengthEstimator` function may return an integer as the percentage where 100% is strong enough, and 0% is too weak.

## Bars
```jsx
<Bars.Root strengthEstimator={myCustomPasswordStrengthEstimator}>
   <Bars.Bar />
   <Bars.Bar />
   <Bars.Bar />
   <Bars.Bar />
   <Bars.Bar />
</Bars.Root>
```

Here, my `myCustomPasswordStrengthEstimator` function may return an integer as the percentage where 100% is strong enough, and 0% is too weak. We could get the result and divide it by the number of `<Bars.Bar />`'s to calculate which bars should be filled and which should not (maybe using `Math.floor`).

What do you think?

Reply:
At the moment the prototypes for those are more low-level so you could wire them to whatever rules you want/have.
However, I can absolutely see some built-in way to provide rules too. We have already explored password strength validation which could check solid strength using https://github.com/dropbox/zxcvbn for example, and also check password breach databases.

Reply:
Got it! Sounds good!